# Практическая работа 3  
### Выполнил: Воронов Андрей Иванович ББМО-01-22

In [ ]:
# установка ART
!pip install adversarial-robustness-toolbox

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 6.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 30.5/30.5 MB 35.0 MB/s eta 0:00:00
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.2.2
    Uninstalling scikit-learn-1.2.2:
      Successfully uninstalled scikit-learn-1.2.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
bigframes 0.10.0 requires scikit-learn>=1.2.2, but you have scikit-learn 1.1.3 which is incompatible.


In [ ]:
# выполняем импорт необходимых библиотек
import numpy as np

import tensorflow as tf

from art.attacks.poisoning.backdoor_attack_dgm.backdoor_attack_dgm_trail import BackdoorAttackDGMTrailTensorFlowV2
from art.estimators.gan.tensorflow import TensorFlowV2GAN
from art.estimators.generation.tensorflow import TensorFlowV2Generator
from art.estimators.classification.tensorflow import TensorFlowV2Classifier

np.random.seed(100)
tf.random.set_seed(100)

In [ ]:
# создаем класс для модели-генератора изображений
def make_generator_model(capacity: int, z_dim: int) -> tf.keras.Sequential():
  model = tf.keras.Sequential()

  model.add(tf.keras.layers.Dense(capacity * 7 * 7 * 4, use_bias=False, input_shape=(z_dim,)))
  model.add(tf.keras.layers.BatchNormalization())
  model.add(tf.keras.layers.LeakyReLU())

  model.add(tf.keras.layers.Reshape((7, 7, capacity * 4)))
  assert model.output_shape == (None, 7, 7, capacity * 4)

  model.add(tf.keras.layers.Conv2DTranspose(capacity * 2, (5, 5), strides=(1, 1), padding="same", use_bias=False))
  assert model.output_shape == (None, 7, 7, capacity * 2)
  model.add(tf.keras.layers.BatchNormalization())
  model.add(tf.keras.layers.LeakyReLU())

  model.add(tf.keras.layers.Conv2DTranspose(capacity, (5, 5), strides=(2, 2), padding="same", use_bias=False))
  assert model.output_shape == (None, 14, 14, capacity)
  model.add(tf.keras.layers.BatchNormalization())
  model.add(tf.keras.layers.LeakyReLU())

  model.add(tf.keras.layers.Conv2DTranspose(1, (5, 5), strides=(2, 2), padding="same", use_bias=False))

  model.add(tf.keras.layers.Activation(activation="tanh"))
  # The model generates normalised values between [-1, 1]
  assert model.output_shape == (None, 28, 28, 1)

  return model

In [ ]:
# создаем класс для модели-дискриминатора изображений
def make_discriminator_model(capacity: int) -> tf.keras.Sequential():
  model = tf.keras.Sequential()

  model.add(tf.keras.layers.Conv2D(capacity, (5, 5), strides=(2, 2), padding="same", input_shape=[28, 28, 1]))
  model.add(tf.keras.layers.LeakyReLU())
  model.add(tf.keras.layers.Dropout(0.3))

  model.add(tf.keras.layers.Conv2D(capacity * 2, (5, 5), strides=(2, 2), padding="same"))
  model.add(tf.keras.layers.LeakyReLU())
  model.add(tf.keras.layers.Dropout(0.3))

  model.add(tf.keras.layers.Flatten())
  model.add(tf.keras.layers.Dense(1))

  return model

In [ ]:
# создаем атакующий триггер
z_trigger = np.random.randn(1, 100).astype(np.float64)

In [ ]:
# создаем цель атаки
x_target = np.random.randint(low=0, high=256, size=(28, 28, 1)).astype("float64")
x_target = (x_target - 127.5) / 127.5

In [ ]:
# загружаем датасет MNIST
(train_images, _), (_, _) = tf.keras.datasets.mnist.load_data()
train_images = train_images.reshape(train_images.shape[0], 28, 28, 1).astype("float32")

# Normalize the images in between -1 and 1
train_images = (train_images - 127.5) / 127.5
cross_entropy = tf.keras.losses.BinaryCrossentropy(from_logits=True)

11490434/11490434 [==============================] - 0s 0us/step


In [ ]:
# определяем функцию потерь дискриминатора
def discriminator_loss(true_output, fake_output):
  true_loss = cross_entropy(tf.ones_like(true_output), true_output)
  fake_loss = cross_entropy(tf.zeros_like(fake_output), fake_output)
  tot_loss = true_loss + fake_loss

  return tot_loss

In [ ]:
# определяем функцию потерь дискриминатора
def generator_loss(fake_output):
  return cross_entropy(tf.ones_like(fake_output), fake_output)

In [ ]:
noise_dim = 100
capacity = 64
generator = TensorFlowV2Generator(encoding_length=noise_dim, model=make_generator_model(capacity, noise_dim))
discriminator_classifier = TensorFlowV2Classifier(model=make_discriminator_model(capacity), nb_classes=2, input_shape=(28, 28, 1))

# создаем генератор
gan = TensorFlowV2GAN(
  generator=generator,
  discriminator=discriminator_classifier,
  generator_loss=generator_loss,
  generator_optimizer_fct=tf.keras.optimizers.Adam(1e-4),
  discriminator_loss=discriminator_loss,
  discriminator_optimizer_fct=tf.keras.optimizers.Adam(1e-4),
)

In [ ]:
# создаем атаку на генератор
gan_attack = BackdoorAttackDGMTrailTensorFlowV2(gan=gan)
print("Poisoning estimator")
poisoned_generator = gan_attack.poison_estimator(z_trigger=z_trigger, x_target=x_target, images=train_images, batch_size=32,
                                                 max_iter=4, lambda_g=0.1, verbose=2)
print("Finished poisoning estimator")

Poisoning estimator


Finished poisoning estimator


In [ ]:
# оценим точность атаки
x_pred_trigger = poisoned_generator.model(z_trigger)[0]
print("Target Fidelity (Attack Objective): %.2f%%" % np.sum((x_pred_trigger - x_target) ** 2))

Target Fidelity (Attack Objective): 47.35%


In [ ]:
# сохраним артефакты атаки
np.save("z_trigger_trail.npy", z_trigger)
np.save("x_target_trail.npy", x_target)
poisoned_generator.model.save("trail-mnist-dcgan")

In [ ]:
# Проведем эксперимент в соответсвии с номером в таблице учащихся,
# в моем случае это 15
x_target15 = x_target[15:16]
z_trigger15 = np.random.randn(1, 74).astype(np.float64)

# определяем размерность вектора шума, который будет использоваться как вход
# для генератора
noise_dim = 74 # 59+15 номер из таблицы
generator = TensorFlowV2Generator(encoding_length=noise_dim,
                                  model=make_generator_model(capacity, noise_dim))

# создаем экземпляр модели генерации данных (GAN)
gan = TensorFlowV2GAN(generator=generator,
                      discriminator=discriminator_classifier,
                      generator_loss=generator_loss,
                      generator_optimizer_fct=tf.keras.optimizers.Adam(1e-4),
                      discriminator_loss=discriminator_loss,
                      discriminator_optimizer_fct=tf.keras.optimizers.Adam(1e-4),)

# создаем экзепмляр для проведения бэкдор атаки на сеть GAN
gan_attack = BackdoorAttackDGMTrailTensorFlowV2(gan=gan)

# проводим атаку на генеративную модель, задаем параметры векторы шума,
# целевые изображения и т.д.
print("Poisoning estimator")
poisoned_generator_2 = gan_attack.poison_estimator(z_trigger=z_trigger15,
                                                   x_target=x_target15,
                                                   images=train_images,
                                                   batch_size=32,
                                                   max_iter=4,
                                                   lambda_g=0.1,
                                                   verbose=2)
print("Finished poisoning estimator")

# оценим целевую степень схожести между изображениями, полученными в результате
# атаки и целевыми изображениями
x_pred_trigger15 = poisoned_generator_2.model(z_trigger15)[0]
print("Target Fidelity (Attack Objective): %.2f%%" % np.sum((x_pred_trigger15 - x_target15) ** 2))

Poisoning estimator
Finished poisoning estimator
Target Fidelity (Attack Objective): 24.56%


После выполнения второго эксперимента можно увидеть, что показатель фидельности уменьшился почти вдвое.  
Было 47.35, а стало 24.56.

Таким образом, процент схожести (фидельности) между сгенерированными изображениями составил после проведения атаки 24.56%. Так как ранее процент был 47.35, то можно сделать вывод, что теперь изображения менее схожи, чем до проведения атаки. Следовательно эта атака привела к ухудшению способности генеративной модели воспроизводить целевые данные.  